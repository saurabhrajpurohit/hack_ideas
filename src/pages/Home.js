import React, { useEffect } from "react";
import { Add, ArrowCircleDownTwoTone, ArrowCircleUpTwoTone } from "@mui/icons-material";
import { Box, Alert, Button, Container, Grid, LinearProgress, Stack, Typography } from "@mui/material";
import Hack from "../components/Hack";
import { HackathonService } from "../service/Service";
import { useNavigate } from "react-router";
import { useSelector } from "react-redux";

export default function Home() {
    const navigate = useNavigate();
    const user = useSelector(state => state.user);
    const [hackathons, setHackathon] = React.useState([]);
    const [loading, setLoading] = React.useState(true);
    const [error, setError] = React.useState("");
    const [sort, setSort] = React.useState({
        field: "CreationDate",
        order: "desc"
    });

    useEffect(() => {
        if (user) {
            getHackathons();
        } else {
            navigate("/login");
        }
    }, [user, navigate]);

    function getHackathons() {
        setLoading(true);
        HackathonService.getAll().then(data => {
            setHackathon(data);
            setLoading(false);
        }).catch(err => {
            console.error(err);
            setError("We are unable to fetch the hackathons at this time. Please try again later.");
            setLoading(false);
        });
    };

    function upVote(upvote, hackathon) {
        if (upvote) {
            hackathon.Votes = [...hackathon.Votes, user.EmployeeID]
        } else {
            hackathon.Votes = hackathon.Votes.filter(vote => vote !== user.EmployeeID);
        }
        setLoading(true);
        HackathonService.update(hackathon.id, hackathon).then(data => {
            getHackathons();
            setLoading(false);
        }).catch(err => {
            console.error(err);
            setError("We are unable to update the hackathon at this time. Please try again later.");
            setLoading(false);
        });
    };

    function handleSort(event) {
        setSort({
            field: event.target.name,
            order: sort.order === "asc" ? "desc" : "asc"
        });
    };

    function getSortedData(hackathons = [], sort) {
        if (sort.field === "Votes") {
            return hackathons.sort((a, b) => {
                return sort.order === "desc" ? a.Votes.length - b.Votes.length : b.Votes.length - a.Votes.length;
            });
        } else {
            return hackathons.sort((a, b) => {
                return sort.order === "desc" ? a[sort.field] - b[sort.field] : b[sort.field] - a[sort.field];
            });
        }
    };

    return (
        <Container data-testid="home-screen">
            <Stack direction="row" alignItems="center" justifyContent="space-between">
                <Typography variant="h5" gutterBottom>
                    Hack Ideas <br />
                    <Typography variant="caption" gutterBottom>
                        Below are the challenges that are currently available.
                    </Typography>
                </Typography>
                <Stack direction="row" spacing={1} sx={{ my: 1 }}>
                    <Button
                        name="Votes"
                        onClick={handleSort}
                        variant="outlined"
                        endIcon={
                            sort.field === "Votes" && (sort.order === "asc"
                                ? <ArrowCircleDownTwoTone />
                                : <ArrowCircleUpTwoTone />)
                        }
                    >
                        Vote
                    </Button>
                    <Button
                        name="CreationDate"
                        onClick={handleSort}
                        variant="outlined"
                        endIcon={
                            sort.field === "CreationDate" && (sort.order === "asc"
                                ? <ArrowCircleDownTwoTone />
                                : <ArrowCircleUpTwoTone />)
                        }
                    >
                        Latest
                    </Button>
                    <Button onClick={() => navigate("/hack/add")} variant="outlined" startIcon={<Add />} color="primary">Add New</Button>
                </Stack>
            </Stack>
            {error && <Box sx={{ width: '100%' }} pb={1}>
                <Alert severity="error">{error}</Alert>
            </Box>}
            {loading && <Box sx={{ width: '100%' }} pb={1}>
                <LinearProgress />
            </Box>}
            <Grid container spacing={3}>
                {getSortedData(hackathons, sort).map((hackathon) => (
                    <Grid key={hackathon.id} item xs={12}>
                        <Hack hackathon={hackathon} handleClick={(upvote) => upVote(upvote, hackathon)} userId={user.EmployeeID} />
                    </Grid>
                ))}
            </Grid>
        </Container>
    )
}