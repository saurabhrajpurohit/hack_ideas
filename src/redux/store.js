import { createStore } from "redux";
import { SET_USER } from "../utils/constants";

const userState = {};

const userReducer = (state = userState, action) => {
    switch (action.type) {
        case SET_USER:
            return {
                ...state,
                user: action.value
            }
        default:
            return state;
    }
};

export default createStore(userReducer);

