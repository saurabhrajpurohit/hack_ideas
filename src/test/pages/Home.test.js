import { fireEvent, render, screen } from '@testing-library/react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import Home from '../../pages/Home';
import store from '../../redux/store';

test('ui to be rendered correctly', () => {
    render(
        <Provider store={store}>
            <BrowserRouter>
                <Home />
            </BrowserRouter>
        </Provider>
    );
    const element = screen.getByTestId('home-screen');
    expect(element).toMatchSnapshot();
});
