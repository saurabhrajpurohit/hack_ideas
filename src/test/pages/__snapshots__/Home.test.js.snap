// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui to be rendered correctly 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
  data-testid="home-screen"
>
  <div
    class="css-1jv2z0n-MuiStack-root"
  >
    <h5
      class="MuiTypography-root MuiTypography-h5 MuiTypography-gutterBottom css-h93ljk-MuiTypography-root"
    >
      Hack Ideas 
      <br />
      <span
        class="MuiTypography-root MuiTypography-caption MuiTypography-gutterBottom css-1hleqvg-MuiTypography-root"
      >
        Below are the challenges that are currently available.
      </span>
    </h5>
    <div
      class="css-vxcp3-MuiStack-root"
    >
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        name="Votes"
        tabindex="0"
        type="button"
      >
        Vote
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        name="CreationDate"
        tabindex="0"
        type="button"
      >
        Latest
        <span
          class="MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="ArrowCircleUpTwoToneIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 20c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m-1-8v4h2v-4h3l-4-4-4 4h3z"
              opacity=".3"
            />
            <path
              d="M12 20c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8m0 2c5.52 0 10-4.48 10-10S17.52 2 12 2 2 6.48 2 12s4.48 10 10 10zm-1-10v4h2v-4h3l-4-4-4 4h3z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButtonBase-root  css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="AddIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
          </svg>
        </span>
        Add New
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  
  <div
    class="MuiBox-root css-olfcrv"
  >
    <span
      class="MuiLinearProgress-root MuiLinearProgress-colorPrimary MuiLinearProgress-indeterminate css-eglki6-MuiLinearProgress-root"
      role="progressbar"
    >
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar1Indeterminate css-5ir5xx-MuiLinearProgress-bar1"
      />
      <span
        class="MuiLinearProgress-bar MuiLinearProgress-barColorPrimary MuiLinearProgress-bar2Indeterminate css-1r8wrcl-MuiLinearProgress-bar2"
      />
    </span>
  </div>
  <div
    class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-3 css-zow5z4-MuiGrid-root"
  />
</div>
`;
